x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/normalizer/ShortCircuitingRenormalizer.java

```diff
@@ -51,169 +62,136 @@ public void renormalize(Quantiles quantiles) {
             return;
         }
 
-        // This will throw NPE if quantiles is null, so do it first
-        QuantilesWithLatch quantilesWithLatch = new QuantilesWithLatch(quantiles, new CountDownLatch(1));
         // Needed to ensure work is not added while the tryFinishWork() method is running
-        synchronized (quantilesDeque) {
-            // Must add to latchDeque before quantilesDeque
-            latchDeque.addLast(quantilesWithLatch.getLatch());
-            quantilesDeque.addLast(quantilesWithLatch);
-            executorService.submit(() -> doRenormalizations());
+        synchronized (this) {
+            latestQuantilesHolder = (latestQuantilesHolder == null)
+                ? new AugmentedQuantiles(quantiles, null, new CountDownLatch(1))
+                : new AugmentedQuantiles(quantiles, latestQuantilesHolder.getEvictedTimestamp(), latestQuantilesHolder.getLatch());
+            // Don't start a thread if another normalization thread is still working. The existing thread will
+            // do this normalization when it finishes its current one. This means we serialise normalizations
+            // without hogging threads or queuing up many large quantiles documents.
+            if (tryStartWork()) {
+                executorService.submit(this::doRenormalizations);
+            }
         }
     }
 
     @Override
-    public void waitUntilIdle() {
-        try {
-            // We cannot tolerate more than one thread running this loop at any time,
-            // but need a different lock to the other synchronized parts of the code
-            synchronized (latchDeque) {
-                for (CountDownLatch latchToAwait = latchDeque.pollFirst(); latchToAwait != null; latchToAwait = latchDeque.pollFirst()) {
-                    latchToAwait.await();
-                }
+    public void waitUntilIdle() throws InterruptedException {
+        CountDownLatch latch;
+        do {
+            // The first bit waits for any not-yet-started renormalization to complete.
+            synchronized (this) {
+                latch = (latestQuantilesHolder != null) ? latestQuantilesHolder.getLatch() : null;
             }
-        } catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
-            throw new RuntimeException(e);
-        }
+            if (latch != null) {
+                latch.await();
+            }
+            // This next bit waits for any thread that's been started to run doRenormalizations() to exit the loop in that method.
+            // If no doRenormalizations() thread is running then we'll wait for the previous phase, and a call to do that should
+            // return immediately.
+            int phaseToWaitFor;
+            synchronized (this) {
+                phaseToWaitFor = phaser.getPhase() - 1 + phaser.getUnarrivedParties();
+            }
+            phaser.awaitAdvanceInterruptibly(phaseToWaitFor);
+        } while (latch != null);
     }
```

Comment:
EDIT: Below was my previous comment, but thinking about it more, I think the loop is still required as we want to allow new normalizes to come in and be executed. We want to make sure all normalization is idle, including handling the case when new ones coming after the current thread stops executing.

------- my old comment for context --------

Does this still need to be in a `do {} while(latch != null)` loop? It seems that the `phaser` is doing the required work here by waiting for the executing thread to finish. 

If the latch is not null that means:
 - We wait for the latch to finish (current quantiles)
 - Wait for the `doRenormalizations` to finish, which may do more than one as `latestQuantilesHolder` can be updated
 - Executor thread completes (meaning all passed quantiles are finished)
 - We go back through the loop again (which isn't synchronized, so another `normalize` call could happen)
 - See the new latch (or maybe not?)
 - go through it again

Reply:
> We want to make sure all normalization is idle, including handling the case when new ones coming after the current thread stops executing.

Yes, that's it.